### PR TITLE
Add package code coverage check

### PR DESCRIPTION
**Pull request by yutongz**
[https://github.com/istio/istio-testing/issues/135](#135)

@sebastienvas

### CODE REVIEW

File: bin/pkg_cc_check.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
+#!/usr/bin/env python
+
+import sys
+
+coverage = {}
+failed_pkgs = []
+
+try :
+    with open('cc_report', 'r') as cc_report :
```
Comment (sebastienvas) [inline bin/pkg_cc_check.py]: Please add a flag to pass in the report file instead of hard coding it.

File: bin/pkg_cc_check.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+            parts = report.split('\t')
+            if (len(parts) > 2) and (parts[0] != "?   ") :
+                words = parts[-1].split(' ')
+                if len(words) > 2 :
+                    coverage[parts[1]] = words[1][:-1]
+except IOError :
+    sys.exit("Failed to read cc_report.")
+
+try :
+    with open('cc_requirements', 'r') as cc_requirements :
```
Comment (sebastienvas) [inline bin/pkg_cc_check.py]: Same comment here. Please pass in a flag.

Comment (sebastienvas) [inline bin/pkg_cc_check.py]: Also, create a class such that you can test it with fake requirement data.

File: bin/pkg_cc_check.py (hunk 1 above)
Comment (yutongz) [inline bin/pkg_cc_check.py]: Done

File: bin/pkg_cc_check.py (hunk 2 above)
Comment (yutongz) [inline bin/pkg_cc_check.py]: Done

File: bin/pkgCheck/pkg_cc_check.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+            print ("Following package(s) failed to satisfied requirements:\n \
+            Package Name    Actual Coverage    Requirement")
+            for pkg in failed_pkgs:
+                print pkg,
+            print ("Failed to satisfied package coverage requirement(s)")
+            return 2
+
+
+if __name__ == "__main__":
+    import sys
```
Comment (sebastienvas) [inline bin/pkgCheck/pkg_cc_check.py]: Please move this import on top of the file.

File: Jenkinsfile [hunk 4]
```diff
@@ -47,7 +47,9 @@ def presubmit(gitUtils, bazel, utils) {
       bazel.test('//...')
     }
     stage('Code Coverage') {
-      sh('bin/codecov.sh')
+      sh('bin/codecov.sh > cc_report')
+      sh('cat cc_report')
+      sh('bin/pkgCheck/pkg_cc_check.py cc_report codecov.info')
```
Comment (sebastienvas) [inline Jenkinsfile]: Dont you need to bazel build it first ?

File: bin/pkgCheck/pkg_cc_check.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
+#!/usr/bin/env python
+#This script checks if every package satisfied package code coverage requirment.
+
+helpinfo = "Usage: pkg_cc_check [cc report file] [cc requirement file]"
```
Comment (sebastienvas) [inline bin/pkgCheck/pkg_cc_check.py]: Please capitalize this global variable

Comment (yutongz) [inline bin/pkgCheck/pkg_cc_check.py]: Done

File: Jenkinsfile (hunk 4 above)
Comment (yutongz) [inline Jenkinsfile]: Done

**Comment by sebastienvas** [comment]
@geeknoid Would this work for you related to https://github.com/istio/istio-testing/issues/135

File: bin/pkgCheck/pkg_cc_check.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+        try:
+            with open(self.cc_requirements, 'r') as cc_requirements_lines:
+                for line in cc_requirements_lines:
+                    self.compRequire(line, failed_pkgs)
+        except IOError as e:
+            print ("Failed to read %s." %(self.cc_requirements), e)
+            return 1
+
+        if not failed_pkgs:
+            print ("All packages passed code coverage requirement!")
```
Comment (geeknoid) [inline bin/pkgCheck/pkg_cc_check.py]: requirements

File: bin/pkgCheck/pkg_cc_check.py [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+            return 0
+        else:
+            print ("Following package(s) failed to satisfied requirements:\n \
```
Comment (geeknoid) [inline bin/pkgCheck/pkg_cc_check.py]: satisfied -> meet

File: bin/pkgCheck/pkg_cc_check.py [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+            Package Name    Actual Coverage    Requirement")
+            for pkg in failed_pkgs:
+                print pkg,
+            print ("Failed to satisfied package coverage requirement(s)")
```
Comment (geeknoid) [inline bin/pkgCheck/pkg_cc_check.py]: to satisfy package code coverage

File: bin/pkgCheck/pkg_cc_check.py [hunk 9]
```diff
@@ -0,0 +1,66 @@
+#!/usr/bin/env python
+#This script checks if every package satisfied package code coverage requirment.
```
Comment (geeknoid) [inline bin/pkgCheck/pkg_cc_check.py]: satisfies
requirements

File: codecov.info [hunk 10]
```diff
@@ -0,0 +1,4 @@
+istio.io/manager/cmd/istioctl   66.4
```
Comment (geeknoid) [inline codecov.info]: How about calling this file codecov.requirements

File: bin/pkgCheck/BUILD [hunk 11]
```diff
@@ -0,0 +1,11 @@
+py_binary(
```
Comment (geeknoid) [inline bin/pkgCheck/BUILD]: Maybe calling this CovCheck would be more descriptive?

Or are you thinking that this tool can also eventually grow up and be used to test perf baselines for packages too?

**Comment by geeknoid** [comment]
How can we share this kind of stuff across repos? It'd be sad to write this tool & infrastructure, and have to copy/paste into the mixer repo and elsewhere.

Do we need a central tool report that is vendored into the other repos?

**Comment by sebastienvas** [comment]
We do istio/toolbox. 
Yutong started working on it, but we got some push back from the team. The goal is to centralize all the tools and check that we do such that all modules have a common experience, but since most of the script are still changing we delayed this.

File: bin/pkgCheck/BUILD (hunk 11 above)
Comment (yutongz) [inline bin/pkgCheck/BUILD]: Could you maybe give me more information, what do you mean by descriptive?

File: WORKSPACE [hunk 12]
```diff
@@ -8,6 +8,12 @@ git_repository(
 
 load("@io_bazel_rules_go//go:def.bzl", "go_repositories", "new_go_repository")
 
+git_repository(
+    name = "ToolBox",
```
Comment (sebastienvas) [inline WORKSPACE]: Please follow the same naming convention as other deps.

File: bin/toolbox/presubmit/package_coverage_check.py [hunk 13]
```diff
@@ -0,0 +1,16 @@
+from ToolBox.pkgCheck.pkg_cc_check import *
```
Comment (sebastienvas) [inline bin/toolbox/presubmit/package_coverage_check.py]: from ToolBox.pkgCheck import pkg_cc_check
https://google.github.io/styleguide/pyguide.html#Imports

**Pull request merged**